Replace oasis-node with oasis in the Run node part (#588)
oasisprotocol/docs: docs/node/run-your-node/validator-node.mdx

# Validator Node

This guide will walk you through the process of setting up your **validator
node** for the Oasis Network either on Mainnet or Testnet. It is designed for
individuals who have basic understanding of the command line environment.

We will be using two separate physical machines for deployment:
- your local system, referred to as `localhost`,
- a remote `server` which will function as an Oasis node.

The guide consists of the following steps:

1. On the `localhost`, we will use [Oasis CLI] to [Initialize your
Entity](#initializing-entity) which is essential for deploying nodes on the
network. To ensure the security of these private keys, we strongly recommend

> to either isolate the localhost from any network or internet connectivity, or

Reviewer: ```suggestion
to either isolate the `localhost` from any network or internet connectivity, or
```

use the [Hardware Security Modules (HSM)] as a secure storage, for example

> the [Ledger wallet].

Reviewer: ```suggestion
use a [Hardware Security Module (HSM)], such as the [Ledger wallet].
```
Maybe?

I also noticed that here we refer to it as HSM, elsewhere as hardware wallet. To keep it simple we should just omit the HSM?

Author: Hardware wallet makes more sense in our docs context.

[Oasis CLI]: ../../general/manage-tokens/cli/README.md
[Hardware Security Modules (HSM)]:
  https://en.wikipedia.org/wiki/Hardware_security_module
[Ledger wallet]:
  ../../../general/manage-tokens/holding-rose-tokens/ledger-wallet

2. After the entity has been created, we will move over to the `server` and
[Start the Oasis Node](#starting-the-oasis-node). The server needs to meet
the hardware requirements and have access to the internet.

[create your entity]: #creating-your-entity
[set up the server in order to start the Oasis node]:
  #running-an-oasis-node-on-the-server

> 3. Finally, we will [stake assets to your entity, register it on the network and

Reviewer: ```suggestion
3. Finally, we will [stake assets to your entity, register it on the network, and
```
Do we follow "Oxford-comma" in our docs? https://www.grammarbook.com/blog/commas/oxford-comma/

Author: Yeah, we do.

associate it with the unique ID](#staking-and-registering) of the
Oasis Node instance running on your server.

## Prerequisites

Before proceeding with this guide, ensure that you have completed the steps
outlined in the [Prerequisites] chapter so that:

* your system meets the [hardware requirements],

> * you have the [Oasis Node binary] installed on your systems,

Reviewer: ```suggestion
* you have the [Oasis CLI binary] installed on your `localhost`,
* you have the [Oasis Node binary] installed on your `server`,
```

structure in mind. Nonetheless, feel free to reorganize and rename directories
as needed to accommodate your preferences.

### Add Entity Account to Oasis CLI

An entity is critical to operating nodes on the network as it controls the stake
attached to a given individual or organization on the network. The entity is
represented as a consensus-layer account using the Ed25519 encryption scheme.
To protect your entity private key, we strongly recommend using a hardware

> wallet such as [Ledger].

Reviewer: ```suggestion
To protect your entity private key, we strongly recommend using a [Hardware Security Module (HSM)],
such as [Ledger].
```

We will be using [Oasis CLI] to initialize the entity and later stake our assets
and register the entity on the network. If you haven't already, go ahead and
install it.

Oasis CLI stores either your entity private key encrypted inside a file or a
reference to an account whose keypair is stored on your hardware wallet.

:::danger

If you really need to use the file-based wallet using another
[offline/air-gapped machine] for this purpose is highly recommended. Gaining
access to the entity private key can compromise your tokens and the network
security through proposing and signing malicious governance transactions.

:::

Now add a new entity account to Oasis CLI. This can be done in one of the following ways:

- Create an account entry in Oasis CLI, but use your Ledger device to store
  the actual keypair to sign the transactions by executing
  [`oasis wallet create`] and passing the `--kind ledger` flag.
- Import your existing `entity.pem` into Oasis CLI by executing

>   [`oasis wallet import-file`] command.

Reviewer 2: Broken link.

>   - Generate a new keypair and store the private key in the encrypted file by

Reviewer 2: Fix indentation.

An entity is critical to operating nodes on the network as it controls the stake
attached to a given individual or organization on the network. The entity is
represented as a consensus-layer account using the Ed25519 encryption scheme.
To protect your entity private key, we strongly recommend using a hardware
wallet such as [Ledger].

We will be using [Oasis CLI] to initialize the entity and later stake our assets
and register the entity on the network. If you haven't already, go ahead and
install it.

Oasis CLI stores either your entity private key encrypted inside a file or a
reference to an account whose keypair is stored on your hardware wallet.

:::danger

If you really need to use the file-based wallet using another
[offline/air-gapped machine] for this purpose is highly recommended. Gaining
access to the entity private key can compromise your tokens and the network
security through proposing and signing malicious governance transactions.

:::

Now add a new entity account to Oasis CLI. This can be done in one of the following ways:

- Create an account entry in Oasis CLI, but use your Ledger device to store
  the actual keypair to sign the transactions by executing

>   [`oasis wallet create`] and passing the `--kind ledger` flag.

Reviewer: Change to an example command for the lazy?

- Import your existing `entity.pem` into Oasis CLI by executing
  [`oasis wallet import-file`] command.
- Generate a new keypair and store the private key in the encrypted file by

>   executing [`oasis wallet create`] and passing the `--kind file` flag.

Reviewer: Change to an example command for the lazy?

In the remainder of the chapter we will assume that you named your entity
account as **`my_entity`**.

### Write the Entity Descriptor File

> Begin by creating a directory named `/localhostdir` with the `entity`

Reviewer: ```suggestion
On `localhost` we begin by creating a directory named `/localhostdir` with the `entity`
```
I would suggest each section starts by mentioning where the commands should be run.

subdirectory that will contain the entity file descriptor.

```shell
mkdir -p /localhostdir/entity
```

Create a JSON file containing the **public key** of your entity by executing
[`oasis account entity init`] and store it as `entity.json`, for example:

```shell

> oasis account entity init /localhost/entity/entity.json --account my_entity

Reviewer: ```suggestion
oasis account entity init /localhostdir/entity/entity.json --account my_entity
```

```

Now, we can move on to configuring our Oasis node with the freshly generated
`entity.json`.

[Ledger]: ../../general/manage-tokens/holding-rose-tokens/ledger-wallet.md
[offline/air-gapped machine]: https://en.wikipedia.org/wiki/Air_gap_\(networking\)
[`oasis wallet create`]: ../../general/manage-tokens/cli/wallet.md#create
[`oasis wallet import`]: ../../general/manage-tokens/cli/wallet.md#import
[`oasis wallet import-file`]: ../../general/manage-tokens/cli/wallet.md#import-file
[`oasis account entity init`]: ../../general/manage-tokens/cli/account.md#entity-init

## Starting the Oasis Node

> ### Setting up the Oasis Node's Working Directory

Reviewer: This section is too bloated with useless information targeting newbies... In my opinion validator node operators should have at least some basic Linux experience and this section can be condensed into something like on https://docs.oasis.io/node/run-your-node/keymanager-node/#prerequisites.

Author: I agree, but let's do this in another PR. The main issue IMO is that there are common sections for setting up *any* node (validator, client, paratime, km, sentry, seed etc.). It should be made consistent by moving the directory creation and permissions section somewhere to the Prerequisites folder and referencing it from everywhere.

```shell
mkdir -p /localhostdir/entity
```

Create a JSON file containing the **public key** of your entity by executing
[`oasis account entity init`] and store it as `entity.json`, for example:

```shell
oasis account entity init -o /localhostdir/entity/entity.json --account my_entity
```

Now, we can move on to configuring our Oasis node with the freshly generated
`entity.json`.

[Ledger]: ../../general/manage-tokens/holding-rose-tokens/ledger-wallet.md
[offline/air-gapped machine]: https://en.wikipedia.org/wiki/Air_gap_\(networking\)
[`oasis wallet create`]: ../../general/manage-tokens/cli/wallet.md#create
[`oasis wallet import`]: ../../general/manage-tokens/cli/wallet.md#import
[`oasis wallet import-file`]: ../../general/manage-tokens/cli/wallet.md#import-file
[`oasis account entity init`]: ../../general/manage-tokens/cli/account.md#entity-init

## Starting the Oasis Node

### Setting up the Oasis Node's Working Directory

Before we run the node on the `server` we need to ensure that we have a place to
store necessary files for the node.

:::info

We will reference the working directory on the `server` as `/serverdir`

> throughout the documentation.

Reviewer: ```suggestion
We will reference the working directory on the `server` as `/node`
(with subdirectories `/node/{etc,bin,runtimes,data}`) throughout the
documentation.
```
Actually, this refactoring is a great opportunity to standardize the `server`-side directories to `/node/{etc,bin,runtimes,data}` (but leave `localhost` to use `/localhostdir`). In my opinion `/node` makes a bit more sense, follows the typical pattern of non-packaged software installed into `/opt`, and is already used in newer docs for node operators.

Mapping:
- `/serverdir/etc` --> `/node/etc`
- `/serverdir/etc/genesis.json` --> `/node/etc/genesis.json`
- `/serverdir/node` --> `/node/data`
- `/serverdir/node/entity/entity.json` --> `/node/etc/entity.json`
- `/usr/local/bin`/`~/.local/bin` (undocumented) --> `/node/bin` (a symlink from either of those locations to `/node/bin/oasis-node` would make it behave the same)

Author: https://github.com/oasisprotocol/cli/pull/159

You can make this directory structure by executing the following command:

> ```shell

Reviewer: Are you sure it is a good idea to render code blocks as `shell`, although we write Bash commands?

Author: We should use standard posix shell commands. Do we use bash specifics anywhere, double [[ and ]] for example?

Reviewer: We don't. I was just wondering if shell renders everything nicely.

mkdir -m700 -p /serverdir/{etc,node,node/entity}
```

#### Copying the Public Entity Artifacts from `/localhostdir`

We will also need to have the public entity artifacts from the `/localhostdir`
present on the `server`. Copy the `/localhostdir/entity/entity.json` file on
`localhost` to `/serverdir/node/entity/entity.json` on the `server`.

#### Copying the Genesis File to the server

The latest Genesis file can be found in the Network Parameters page ([Mainnet],
[Testnet]). You should download the latest `genesis.json` file and
copy it to `/serverdir/etc/genesis.json` on the `server`.

[Mainnet]: ../mainnet/README.md
[Testnet]: ../testnet/README.md

#### Configuring the Oasis Node

There are a variety of options available when running an Oasis node. The
following YAML file is a basic configuration for a validator node on the
network.

Before using this configuration you should collect the following information to
replace the  variables present in the configuration file:

> * `{{ external_address }}`: The external IP you used when registering this node.

Reviewer: ```suggestion
* `{{ external_ip }}`: The external/public IP address you used when registering this node.
```
In order to not be confused with other types of addresses (e.g. `seed_node_address`), this one must be an IP address, therefore it would be better to refer to it in this way.

:::info

If you are using a [Sentry Node](sentry-node.md), you should use the public IP
of that machine.

:::

* `{{ seed_node_address }}`: The seed node address in the form `ID@IP:port`.

>   You can find the current Oasis Seed Node address in the [Network Parameters].

Reviewer: We typically link to both, because Testnet and Mainnet Network Parameters differ.

You can make this directory structure by executing the following command:

```shell

> mkdir -m700 -p /serverdir/{etc,node,node/entity}

Reviewer: ```suggestion
mkdir -m700 -p /node/{etc,bin,runtimes,data}
```

```

#### Copying the Public Entity Artifacts from `/localhostdir`

We will also need to have the public entity artifacts from the `/localhostdir`
present on the `server`. Copy the `/localhostdir/entity/entity.json` file on

> `localhost` to `/serverdir/node/entity/entity.json` on the `server`.

Reviewer: ```suggestion
`localhost` to `/node/etc/entity.json` on the `server`.
```
In my opinion `entity.json` is more part of the configuration (it is referenced from `config.yml` in `/node/etc`, like `genesis.json`), than part of the node identity (`*.pem` files in `/node/data`).

Author: I agree. Also, entity.json will probably be replaced by a simple entity_id in the oasis-node config and more automated Oasis CLI for attaching/removing nodes to your entity (ideally without the need for the entity descriptor file).

#### Copying the Genesis File to the server

The latest Genesis file can be found in the Network Parameters page ([Mainnet],
[Testnet]). You should download the latest `genesis.json` file and

> copy it to `/serverdir/etc/genesis.json` on the `server`.

Reviewer: ```suggestion
copy it to `/node/etc/genesis.json` on the `server`.
```

[Mainnet]: ../mainnet/README.md
[Testnet]: ../testnet/README.md

#### Configuring the Oasis Node

There are a variety of options available when running an Oasis node. The
following YAML file is a basic configuration for a validator node on the
network.

Before using this configuration you should collect the following information to

> replace the  variables present in the configuration file:

Reviewer: ```suggestion
replace the variables present in the configuration file:
```

* `{{ external_ip }}`: The external/public IP address you used when registering
  this node.

:::info

If you are using a [Sentry Node](sentry-node.md), you should use the public IP
of that machine.

:::

* `{{ seed_node_address }}`: The seed node address in the form `ID@IP:port`.

  You can find the current Oasis Seed Node address in the Network Parameters
  ([Mainnet], [Testnet]).

> To use this configuration, save it in the `/serverdir/etc/config.yml` file:

Reviewer: ```suggestion
To use this configuration, save it in the `/node/etc/config.yml` file:
```

# Genesis.
genesis:
  # Path to the genesis file for the current version of the network.
  file: /serverdir/etc/genesis.json

# Peer-to-peer.
p2p:
  port: 9200
  registration:
    addresses:
      - "{{ external_address }}:9200"
  # List of seed nodes to connect to.
  # NOTE: You can add additional seed nodes to this list if you want.
  seeds:
    - "{{ seed_node_address }}"

# In order for the node to register itself, the entity.json of the entity
# used to provision the node must be available on the node.
registration:
  entity: /serverdir/node/entity/entity.json
```

#### Starting Oasis Node for the First Time

Execute the Oasis node and pass the config file above by invoking:

```shell
oasis-node --config /serverdir/etc/config.yml
```

After the node starts, it will automatically register itself to the peer-to-peer
network. The following files will be generated in your `/serverdir/node` folder:

* `consensus.pem`: The node's consensus private key. **DO NOT SHARE**
* `consensus_pub.pem`: The node's consensus public key.
* `identity.pem`: The node's identity private key. **DO NOT SHARE**
* `identity_pub.pem`: The node's identity public key.

> * `node_genesis.json`: The node's details if you wish to include this node in

Reviewer 2: I believe this is not generated in this case and should be removed. It was only generated when an explicit `node init` was used before.

Author: Running `oasis-node --config config.yml` actually generates those files on my machine in the data dir. Shouldn't it? I'm using version 23.0.4.
In the logs:
```
{"caller":"init.go:97","consensus_pk":"awqeaPPOzypPSoy4tkTV9BRPvLQ4XN5N2nba8tUtetM=","level":"info","module":"oasis-node","msg":"loaded/generated node identity","node_pk":"vvD3uf/9PyXIl7kPpvFxVw6HEA9IJXvE1QJ++OlkaXk=","p2p_pk":"f2vUedQX/yPSQ5qlDrakbDKeH2o2x1BygCYobkqK0Kc=","tls_pk":"9cwTl3qgJ7OrZnzoxheOcTElHYYFngAzBZBwBI8mgqY=","ts":"2023-11-15T10:49:14.446168035Z"}
```

Reviewer: If I remember correctly, in all versions os Oasis Node these files were generated if they did not exist.

Reviewer 2: I was referring specifically to `node_genesis.json`, not the other key files.

        - "{{ seed_node_address }}"
```

#### Starting Oasis Node for the First Time

Execute the Oasis node and pass the config file above by invoking:

```shell
oasis-node --config /serverdir/etc/config.yml
```

After the node starts, it will automatically register itself to the peer-to-peer
network. The following files will be generated in your `/serverdir/node` folder:

* `consensus.pem`: The node's consensus private key. **DO NOT SHARE**
* `consensus_pub.pem`: The node's consensus public key.
* `identity.pem`: The node's identity private key. **DO NOT SHARE**
* `identity_pub.pem`: The node's identity public key.
* `node_genesis.json`: The node's details if you wish to include this node in
  the genesis file of the network.
* `p2p.pem`: The node's private key for libp2p. **DO NOT SHARE**
* `p2p_pub.pem`: The node's public key for libp2p.
* `sentry_client_tls_identity.pem`: The node's TLS private key for communicating
  with sentry nodes. **DO NOT SHARE**
* `sentry_client_tls_identity_cert.pem`: The node's TLS certificate for
  communicating with sentry nodes.

> #### Ensuring Proper Permissions

Reviewer: As already mentioned, this can be condensed into https://docs.oasis.io/node/run-your-node/keymanager-node/#prerequisites. If you do not have proper permission Oasis Node will fail to start and complain in the logs.

Author: See https://github.com/oasisprotocol/docs/pull/588#discussion_r1397689481

        - "{{ seed_node_address }}"
```

#### Starting Oasis Node for the First Time

Execute the Oasis node and pass the config file above by invoking:

```shell
oasis-node --config /serverdir/etc/config.yml
```

After the node starts, it will automatically register itself to the peer-to-peer
network. The following files will be generated in your `/serverdir/node` folder:

* `consensus.pem`: The node's consensus private key. **DO NOT SHARE**
* `consensus_pub.pem`: The node's consensus public key.
* `identity.pem`: The node's identity private key. **DO NOT SHARE**
* `identity_pub.pem`: The node's identity public key.
* `p2p.pem`: The node's private key for libp2p. **DO NOT SHARE**
* `p2p_pub.pem`: The node's public key for libp2p.
* `sentry_client_tls_identity.pem`: The node's TLS private key for communicating
  with sentry nodes. **DO NOT SHARE**
* `sentry_client_tls_identity_cert.pem`: The node's TLS certificate for
  communicating with sentry nodes.

#### Ensuring Proper Permissions

Only the owner of the process that runs the Oasis node should have access to the

> files in the `/serverdir/node` directory. The `oasis-node` binary ensures that

Reviewer: ```suggestion
files in the `/node/data` directory. The `oasis-node` binary ensures that
```

the files used by the node are as least privileged as possible so that you don't
accidentally shoot yourself in the foot while operating a node.

To ensure proper permissions are set, we suggest running the following to remove
all non-owner read/write/execute permissions:

```shell

> chmod -R go-r,go-w,go-x /serverdir

Reviewer: ```suggestion
chmod -R go-r,go-w,go-x /node
```

```

:::info

Just so it's clear, the following permissions are expected by the `oasis-node`
binary:

* `700` for the `/serverdir/node` directory
* `700` for the `/serverdir/node/entity` directory

> * `600` for all `*.pem` files

Reviewer: ```suggestion
* `700` for the `/node/data` directory
* `600` for all `/node/data/*.pem` files
```

:::

### Obtaining the node ID

Now that the Oasis node is running, you can obtain your unique node ID which is
needed in order to associate your node with your entity in the network registry.

```sh

> oasis-node control status -a unix:/serverdir/internal.sock | jq .identity.node

Reviewer: ```suggestion
oasis-node control status -a unix:/serverdir/node/internal.sock | jq .identity.node
```

```

```
"5MsgQwijUlpH9+0Hbyors5jwmx7tTmKMA4c9leV3prI="
```

### Check that your node is synced

Before you can become a validator, you will have to make sure that your node is
synced. To do so call this command on the server:

```shell

> oasis-node control status -a unix:/serverdir/node/internal.sock | jq .consensus.status

Reviewer: `oasis-node control is-synced` still works.

```

If your node is synced, the above command should output:

```
"ready"
```

If your node is not yet synced, you will need to wait before you can move
forward.

## Staking and Registering {#staking-and-registering}

Once you have been funded, you can complete the process of connecting your node
to the network by registering both your entity and your node, as described
below.

### Staking (Escrow) Transaction

The current minimum stake required to register an entity and register a node as
a validator is 200 tokens. We will generate an escrow transaction that
delegates 200 tokens from your entity account on the consensus layer to itself.

```shell
oasis account delegate 200 my_entity --no-paratime --account my_entity
```

If you want to use a different account to stake your tokens from,

> [import][`oasis wallet import`] it to Oasis CLI and use the `--account`

Reviewer: Change to an example command for the lazy?

:::

### Obtaining the node ID

Now that the Oasis node is running, you can obtain your unique node ID which is
needed in order to associate your node with your entity in the network registry.

```sh
oasis-node control status -a unix:/serverdir/internal.sock | jq .identity.node
```

```
"5MsgQwijUlpH9+0Hbyors5jwmx7tTmKMA4c9leV3prI="
```

### Check that your node is synced

Before you can become a validator, you will have to make sure that your node is
synced. To do so call this command on the server:

```shell
oasis-node control status -a unix:/serverdir/node/internal.sock | jq .consensus.status
```

If your node is synced, the above command should output:

```
"ready"
```

If your node is not yet synced, you will need to wait before you can move
forward.

## Staking and Registering {#staking-and-registering}

:::tip

> This step is not necessary if your entity was fully staked at genesis.

Reviewer 2: Remove this as it is no longer relevant.

:::

Once you have been funded, you can complete the process of connecting your node
to the network by registering both your entity and your node, as described
below.

### Staking (Escrow) Transaction

The current minimum stake required to register an entity and register a node as
a validator is 200 tokens. We will generate an escrow transaction that

> delegates 200 tokens from your account on the consensus layer to your own
> entity account.

Reviewer 2: This is a bit confusing/awkward. Maybe something like this:
```suggestion
delegates 200 tokens from your entity account on the consensus layer to itself.
```

```shell
oasis account delegate 200 my_entity --no-paratime --account my_entity
```

If you want to use a different account to stake your tokens from,
[import][`oasis wallet import`] it to Oasis CLI and use the `--account`
parameter to pass it.

### Adding your Node ID to the Entity Descriptor

Now we can register our entity on the network and associate it with the node ID

> obtained in the [section above][#obtaining-the-node-id]. Open the `entity.json`

Reviewer 2: Broken link?

Author: Oops, wrong markdown syntax.

file we initially generated and add the ID inside the `nodes` block. Your
entity descriptor file should now look like this:

```json
{
  "id": "Bx6gOixnxy15tCs09ua5DcKyX9uo2Forb32O6Hyjoc8=",
  "nodes": [
    "5MsgQwijUlpH9+0Hbyors5jwmx7tTmKMA4c9leV3prI="
  ],
  "v": 2
}
```

### Entity Registration

We can submit the fresh entity file descriptor by invoking the
[`oasis account entity register`] command:

```shell

> oasis account entity register entity.json

Reviewer 2: ```suggestion
oasis account entity register entity.json --account my_entity
```
